krabcake/kc_main.c

```diff
@@ -113,14 +113,790 @@ static void kc_post_clo_init(void)
+}
+
+/* FIXME: leaving out tracing of the stored data here for now. */
+static void trace_storeg(ULong guard, Addr addr, SizeT size)
+{
+   VG_(printf)("trace_storeg guard=%llu addr=%08lx size=%lu\n", guard, addr, size);
+}
+
+static void trace_loadg(ULong guard, Addr addr, SizeT size, SizeT widened_size)
+{
+   // VG_(printf)("trace_loadg guard=%lld addr=%08llx size=%llu\n", guard, addr, size);
+}
+
 static void trace_wrtmp_load(Addr addr, SizeT size)
 {
    // VG_(printf)("trace_wrtmp_load addr=%08llx size=%llu\n", addr, size);
 }
 
-static void trace_llsc()
+static void trace_store(Addr addr, ULong data, SizeT size)
+{
+   // VG_(printf)("trace_store addr=%08lx data=%08llx size=%llu\n", addr, data, size);
+}
+
+static void trace_store128(Addr addr, ULong data1, ULong data2, SizeT size)
+{
+   VG_(printf)("trace_store128 addr=%08lx data=[%08llx, %08llx] size=%lu\n", addr, data1, data2, size);
+}
+
+static void trace_store256(Addr addr, ULong data1, ULong data2, ULong data3, ULong data4, SizeT size)
+{
+   VG_(printf)("trace_store128 addr=%08lx data=[%08llx, %08llx, %08llx, %08llx] size=%lu\n", addr, data1, data2, data3, data4, size);
+}
+
+static void trace_llsc(Addr addr)
+{
+   VG_(printf)("trace_llsc addr=%08lx\n", addr);
+}
+
+typedef enum { Orig=1, Tag=2, Stk=3, Vsh=4 } TempKind;
+
+/* Create a new IRTemp of type 'ty' and kind 'kind'.
+   See analogous function in memcheck for ideas of where this might go.
+*/
+static IRTemp newTemp ( IRSB *sbOut, IRType ty, TempKind kind )
+{
+   IRTemp     tmp = newIRTemp(sbOut->tyenv, ty);
+   return tmp;
+}
+
+/* assign value to tmp */
+static inline 
+void assign ( HChar cat, IRSB *sbOut, IRTemp tmp, IRExpr* expr ) {
+   addStmtToIRSB(sbOut, IRStmt_WrTmp(tmp,expr));
+}
+
+#define mkexpr(_tmp)             IRExpr_RdTmp((_tmp))
+
+typedef  IRExpr  IRAtom;
+
+/* Bind the given expression to a new temporary, and return the
+   temporary.  This effectively converts an arbitrary expression into
+   an atom.
+
+   'ty' is the type of 'e' and hence the type that the new temporary
+   needs to be.  But passing it in is redundant, since we can deduce
+   the type merely by inspecting 'e'.  So at least use that fact to
+   assert that the two types agree. */
+static IRAtom* assignNew ( HChar cat, IRSB *sbOut, IRType ty, IRExpr* e )
+{
+   TempKind k;
+   IRTemp   t;
+   IRType   tyE = typeOfIRExpr(sbOut->tyenv, e);
+
+   tl_assert(tyE == ty); /* so 'ty' is redundant (!) */
+   switch (cat) {
+      case 'T': k = Tag;  break;
+      case 'S': k = Stk;  break;
+      case 'V': k = Vsh;  break;
+      case 'C': k = Orig; break; 
+                /* happens when we are making up new "orig"
+                   expressions, for IRCAS handling */
+      default: tl_assert(0);
+   }
+   t = newTemp(sbOut, ty, k);
+   assign(cat, sbOut, t, e);
+   return mkexpr(t);
+}
+
+typedef enum ExprContext {
+   /* These are the recursive occcurences of IExpr in the grammar of its definition */
+   Expr_GetI_Index = 0x1F00,
```

Comment:
where does this number come from? might be good to leave a comment saying what other file it comes from, if there is one (that way it's easy to find later once we forget)

Reply:
Oh, its not coming from another file. That's just so I have a better chance of recognizing it in a debugger.

Reply:
Well, okay, *this* number is not coming from another file.

Reply:
The *sequence* of numbers is coming from me scanning over the occurrences of `=0x` in https://github.com/pnkfelix/krabcake-vg/blob/krabcake-vg/VEX/pub/libvex_ir.h; the last one in the file is this:

https://github.com/pnkfelix/krabcake-vg/blob/45a8f67429aa65395176fbab8e721c46f7c40cc1/VEX/pub/libvex_ir.h#L2873

so I took that, and then bumped the `E` to an `F`

I can write a comment explaining this.